### PR TITLE
Hot Reload of Localised Strings

### PR DESCRIPTION
**Pull request by kuanyi-ng**
## Summary
This is an attempt to implement the "Hot Reload" feature for `stacked_localisation` when the user changes the device's language. (Issue brought up in #126)

## Details
- 4fd5fe0
  - the example app provided in `/example` couldn't be start up due to `ambiguous_import` error as the `Router` class is defined in both `package:example/core/router.gr.dart` and `package:flutter/src/widgets/router.dart`.
  - this is solved by hiding the `Router` class provided in `flutter/material.dart`.
- e1b0f08
  - by adding `WidgetsBindingObserver` functionality to `LocalisationService`, we get access to `didChangeLocales()` and `didChangeAppLifecycleState()` callbacks.
  - there are two different ways of implementation that will load up the new strings (of the new selected device's language) and place them into memory.
  - implementation with `didChangeLocales()`: since we have access to the devices' `locale` from the callback function, we just have to update `_localisedStrings` with `getStringsFromAssets` instead of running `initialise()`.
  ```dart
  @override
  void didChangeLocales(List<Locale> locale) async {
    final currentLocale = locale.first.toString();
    _localisedStrings = await _stringReader.getStringsFromAssets(currentLocale);
  }
  ```
  - implementation with `didChangeAppLifecycleState()`: run `initialise()` everytime the app is brought up from background.
  ```dart
  @override
  void didChangeAppLifecycleState(AppLifecycleState state) async {
    // If the user changes language on the device and come back to the app,
    // it will trigger this function with AppLifecycleState.resumed
    if (state == AppLifecycleState.resumed) {
      await _instance.initialise();
    }
  }
  ```

### Current Progress

As shown in the GIF above, after we change the device's language from en to af
- contents in `HomeView` stayed in en
- contents in the dialog changed to af

From this, we know that `didChangeLocales()` callback ran and updated `_localisedStrings`.
However, `HomeView` is not rerendered after `didChangeLocales()` callback in LocalisationService.
`HomeView` will rerender when we perform Flutter Hot Reload.

❓ Not sure if it's possible to trigger rendering of `HomeView` from LocalisationService after `_localisedString` is updated.

### CODE REVIEW

**Comment by FilledStacks** [comment]
awesome! Sorry for the late checking, i'll review this today after my calls and then i'll merge if it's ready. Dankie!

**Comment by kuanyi-ng** [comment]
It's okay! Let me know if it fits the feature's requirements and if there's any changes I need to make. ✌🏻

**Comment by FilledStacks** [comment]
Just want to double check. The gif shows it not working, is this working?

**Comment by kuanyi-ng** [comment]
LocalisationService reinitialised after the language changed but the changes weren't reflected on HomeView.
But the content in the dialog was shown with the newly selected language. (shown in the GIF)

**Comment by FilledStacks** [comment]
I see. It's definitely an improvement, will need to build it out a bit more to ensure the UI is reloaded as well.

**Review by FilledStacks — APPROVED** [review]

**Comment by kuanyi-ng** [comment]
Thank you for merging. 😃 
I will research on ways to trigger UI re-rendering from Service in this architecture.
I wonder if it's possible to get some resources (documentation to read, video to watch) from you on this topic as I started Flutter development only few months ago. 🙇🏻

**Pull request merged**
